File: Lib/Telerik/Examples/QuickStart/GridView/CustomViews/ReflectiveImagePrimitive.vb
Imports System.Text
Imports Telerik.WinControls.UI
Imports Telerik.WinControls.Primitives

Namespace Telerik.Examples.WinControls.GridView.CustomViews
	Public Class ReflectiveImagePrimitive
		Inherits LightVisualElement
		#Region "Fields"

		Private imagePrimitive As ImagePrimitive
		Private reflectionPrimitive As ReflectionPrimitive

		#End Region

		#Region "Properties"

		Public Overrides Property Image() As Image
			Get
				Return Me.imagePrimitive.Image
			End Get
			Set(ByVal value As Image)
				If Me.imagePrimitive.Image IsNot value Then
					Me.imagePrimitive.Image = value
					If Me.Parent IsNot Nothing Then
						Me.Parent.Invalidate()
					End If
				End If
			End Set
		End Property

		#End Region

		#Region "Initialization"

		Protected Overrides Sub InitializeFields()
			MyBase.InitializeFields()
			Me.StretchVertically = False
			Me.StretchHorizontally = False
		End Sub

		Protected Overrides Sub CreateChildElements()
			MyBase.CreateChildElements()

			Me.imagePrimitive = New ImagePrimitive()
			Me.Children.Add(Me.imagePrimitive)

			Me.reflectionPrimitive = New ReflectionPrimitive(Me.imagePrimitive)
			Me.reflectionPrimitive.ItemReflectionPercentage = 0.5
			Me.Children.Add(Me.reflectionPrimitive)
		End Sub

		#End Region

		#Region "Layout"

		Protected Overrides Overloads Function MeasureOverride(ByVal availableSize As SizeF) As SizeF
			MyBase.MeasureOverride(availableSize)
			Return New SizeF(Me.imagePrimitive.DesiredSize.Width, Me.imagePrimitive.DesiredSize.Height * CSng(Me.reflectionPrimitive.ItemReflectionPercentage))
		End Function

		Protected Overrides Function ArrangeOverride(ByVal finalSize As SizeF) As SizeF
			MyBase.ArrangeOverride(finalSize)

			Me.imagePrimitive.Arrange(New RectangleF(0, 0, finalSize.Width, finalSize.Height / CSng(Me.reflectionPrimitive.ItemReflectionPercentage)))
			Me.reflectionPrimitive.Arrange(New RectangleF(0, Me.imagePrimitive.DesiredSize.Height, finalSize.Width, Me.reflectionPrimitive.DesiredSize.Height / 3f))

			Return finalSize
		End Function

		#End Region
	End Class
End Namespace

File: Lib/Telerik/Examples/QuickStart/GridView/CustomViews/ReflectiveImagePrimitive.cs
using System;
using System.Collections.Generic;
using System.Text;
using Telerik.WinControls.UI;
using Telerik.WinControls.Primitives;
using System.Drawing;

namespace Telerik.Examples.WinControls.GridView.CustomViews
{
    public class ReflectiveImagePrimitive : LightVisualElement
    {
        #region Fields

        private ImagePrimitive imagePrimitive;
        private ReflectionPrimitive reflectionPrimitive;

        #endregion

        #region Properties

        public override Image Image
        {
            get
            {
                return this.imagePrimitive.Image;
            }
            set
            {
                if (this.imagePrimitive.Image != value)
                {
                    this.imagePrimitive.Image = value;
                    if (this.Parent != null)
                    {
                        this.Parent.Invalidate();
                    }
                }
            }
        }

        #endregion

        #region Initialization

        protected override void InitializeFields()
        {
            base.InitializeFields();
            this.StretchVertically = false;
            this.StretchHorizontally = false;
        }

        protected override void CreateChildElements()
        {
            base.CreateChildElements();

            this.imagePrimitive = new ImagePrimitive();
            this.Children.Add(this.imagePrimitive);

            this.reflectionPrimitive = new ReflectionPrimitive(this.imagePrimitive);
            this.reflectionPrimitive.ItemReflectionPercentage = 0.5;
            this.Children.Add(this.reflectionPrimitive);
        }

        #endregion

        #region Layout

        protected override SizeF MeasureOverride(SizeF availableSize)
        {
            base.MeasureOverride(availableSize);
            return new SizeF(this.imagePrimitive.DesiredSize.Width, this.imagePrimitive.DesiredSize.Height * (float)this.reflectionPrimitive.ItemReflectionPercentage);
        }

        protected override SizeF ArrangeOverride(SizeF finalSize)
        {
            base.ArrangeOverride(finalSize);

            this.imagePrimitive.Arrange(new RectangleF(0, 0, finalSize.Width, finalSize.Height / (float)this.reflectionPrimitive.ItemReflectionPercentage));
            this.reflectionPrimitive.Arrange(new RectangleF(0, this.imagePrimitive.DesiredSize.Height, finalSize.Width, this.reflectionPrimitive.DesiredSize.Height / 3f));

            return finalSize;
        }

        #endregion
    }
}
